File: pricing-service-java/src/main/java/com/example/retail/javaapp/cache/CacheKey.java
package com.example.retail.javaapp.cache;

import com.example.retail.contract.PriceQuoteRequest;

public final class CacheKey {

    private CacheKey() {
    }

    public static String from(PriceQuoteRequest request) {
        String coupon = request.couponCode() == null || request.couponCode().isBlank()
                ? "none"
                : request.couponCode().trim().toUpperCase();

        return "quote:%s:%s:%s:%s:%s:%s".formatted(
                sanitize(request.sku()),
                sanitize(request.storeId()),
                request.channel().name(),
                request.customerSegment().name(),
                request.quantity(),
                sanitize(coupon)
        );
    }

    private static String sanitize(String value) {
        return value.replace(":", "_").trim().toUpperCase();
    }
}

File: pricing-service-kotlin/src/main/kotlin/com/example/retail/kotlinapp/cache/CacheKey.kt
package com.example.retail.kotlinapp.cache

import com.example.retail.contract.PriceQuoteRequest

object CacheKey {
    fun from(request: PriceQuoteRequest): String {
        val coupon = request.couponCode()
            ?.takeIf { it.isNotBlank() }
            ?.trim()
            ?.uppercase()
            ?: "NONE"

        return listOf(
            "quote",
            sanitize(request.sku()),
            sanitize(request.storeId()),
            request.channel().name,
            request.customerSegment().name,
            request.quantity().toString(),
            sanitize(coupon),
        ).joinToString(":")
    }

    private fun sanitize(value: String): String = value.replace(":", "_").trim().uppercase()
}

